Import Libaries 

In [2]:
import streamlit as st
# importing libraries
import pandas as pd
import plotly.express as px
import altair as alt

Load the Dataset

In [3]:
# Load data
vehicles = pd.read_csv('vehicles_us.csv')

In [4]:
# Display the first 5 rows of the dataset
vehicles.head()

price  model_year           model  condition  cylinders fuel  odometer  \
0   9400      2011.0          bmw x5       good        6.0  gas  145000.0   
1  25500         NaN      ford f-150       good        6.0  gas   88705.0   
2   5500      2013.0  hyundai sonata   like new        4.0  gas  110000.0   
3   1500      2003.0      ford f-150       fair        8.0  gas       NaN   
4  14900      2017.0    chrysler 200  excellent        4.0  gas   80903.0   

  transmission    type paint_color  is_4wd date_posted  days_listed  
0    automatic     SUV         NaN     1.0  2018-06-23           19  
1    automatic  pickup       white     1.0  2018-10-19           50  
2    automatic   sedan         red     NaN  2019-02-07           79  
3    automatic  pickup         NaN     NaN  2019-03-22            9  
4    automatic   sedan       black     NaN  2019-04-02           28

In [5]:
# Display the last 5 rowso f the dataset
vehicles.tail()

price  model_year           model  condition  cylinders fuel  odometer  \
51520   9249      2013.0   nissan maxima   like new        6.0  gas   88136.0   
51521   2700      2002.0     honda civic    salvage        4.0  gas  181500.0   
51522   3950      2009.0  hyundai sonata  excellent        4.0  gas  128000.0   
51523   7455      2013.0  toyota corolla       good        4.0  gas  139573.0   
51524   6300      2014.0   nissan altima       good        4.0  gas       NaN   

      transmission   type paint_color  is_4wd date_posted  days_listed  
51520    automatic  sedan       black     NaN  2018-10-03           37  
51521    automatic  sedan       white     NaN  2018-11-14           22  
51522    automatic  sedan        blue     NaN  2018-11-15           32  
51523    automatic  sedan       black     NaN  2018-07-02           71  
51524    automatic  sedan         NaN     NaN  2018-06-05           10

Understand the Dataset

In [6]:
# Summary statistics
print("Descriptive statistics:")
vehicles.describe()

Descriptive statistics:


price    model_year     cylinders       odometer   is_4wd  \
count   51525.000000  47906.000000  46265.000000   43633.000000  25572.0   
mean    12132.464920   2009.750470      6.125235  115553.461738      1.0   
std     10040.803015      6.282065      1.660360   65094.611341      0.0   
min         1.000000   1908.000000      3.000000       0.000000      1.0   
25%      5000.000000   2006.000000      4.000000   70000.000000      1.0   
50%      9000.000000   2011.000000      6.000000  113000.000000      1.0   
75%     16839.000000   2014.000000      8.000000  155000.000000      1.0   
max    375000.000000   2019.000000     12.000000  990000.000000      1.0   

       days_listed  
count  51525.00000  
mean      39.55476  
std       28.20427  
min        0.00000  
25%       19.00000  
50%       33.00000  
75%       53.00000  
max      271.00000

In [7]:
# Check the shape of the dataset (rows, columns)
print("Shape:", vehicles.shape)

Shape: (51525, 13)


In [8]:
# Get a summary of the dataset, including data types
print("Info:")
vehicles.info()

Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51525 entries, 0 to 51524
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   price         51525 non-null  int64  
 1   model_year    47906 non-null  float64
 2   model         51525 non-null  object 
 3   condition     51525 non-null  object 
 4   cylinders     46265 non-null  float64
 5   fuel          51525 non-null  object 
 6   odometer      43633 non-null  float64
 7   transmission  51525 non-null  object 
 8   type          51525 non-null  object 
 9   paint_color   42258 non-null  object 
 10  is_4wd        25572 non-null  float64
 11  date_posted   51525 non-null  object 
 12  days_listed   51525 non-null  int64  
dtypes: float64(4), int64(2), object(7)
memory usage: 5.1+ MB


In [9]:
# Check for missing values
print("Missing values:")
vehicles.isnull().sum()

Missing values:


price               0
model_year       3619
model               0
condition           0
cylinders        5260
fuel                0
odometer         7892
transmission        0
type                0
paint_color      9267
is_4wd          25953
date_posted         0
days_listed         0
dtype: int64

Missing Values Overview
(Low Percentage < 5%, High Percentage > 20%)
* model_year: 3619 missing values (likely significant, depending on total rows 51525). 
* cylinders: 5260 missing values (significant)
* odometer: 7892 missing values (significant)
* paint color: 9267 missing values (significant)
* is_4wd: 25953 missing values (very significant)

For this dataset review, the option is to drop missing values, versus imputation (instead of dropping missing values considering filling in with a specific values, mean, median, etc.)

In [10]:
# Keeping track of dropped data
vehicles_shape = vehicles.shape
vehicles_dropped = vehicles.dropna()
print(f"Dropped {vehicles_shape[0] - vehicles_dropped.shape[0]} rows with missing values.")

Dropped 36673 rows with missing values.


In [14]:
vehicles_dropped.head()

price  model_year                     model  condition  cylinders fuel  \
5   14990      2014.0              chrysler 300  excellent        6.0  gas   
7   15990      2013.0               honda pilot  excellent        6.0  gas   
10  19500      2011.0  chevrolet silverado 1500  excellent        8.0  gas   
14  12990      2009.0                 gmc yukon  excellent        8.0  gas   
16  14990      2010.0                  ram 1500  excellent        8.0  gas   

    odometer transmission    type paint_color  is_4wd date_posted  days_listed  
5    57954.0    automatic   sedan       black     1.0  2018-06-20           15  
7   109473.0    automatic     SUV       black     1.0  2019-01-07           68  
10  128413.0    automatic  pickup       black     1.0  2018-09-17           38  
14  132285.0    automatic     SUV       black     1.0  2019-01-31           24  
16  130725.0    automatic  pickup         red     1.0  2018-12-30           13

Descriptive Statistics 

In [11]:
vehicles_dropped.describe()

price    model_year     cylinders       odometer   is_4wd  \
count   14852.000000  14852.000000  14852.000000   14852.000000  14852.0   
mean    15446.921155   2010.052586      6.798950  117713.691287      1.0   
std     11393.676296      5.790532      1.468353   65610.650451      0.0   
min         1.000000   1965.000000      3.000000       0.000000      1.0   
25%      6989.500000   2006.000000      6.000000   71597.500000      1.0   
50%     13500.000000   2011.000000      8.000000  115665.500000      1.0   
75%     21988.000000   2015.000000      8.000000  158378.000000      1.0   
max    375000.000000   2019.000000     12.000000  866000.000000      1.0   

        days_listed  
count  14852.000000  
mean      39.863520  
std       28.073348  
min        0.000000  
25%       19.000000  
50%       34.000000  
75%       54.000000  
max      263.000000

Data Visualization

In [12]:
vehicles_dropped.columns.tolist()

['price',
 'model_year',
 'model',
 'condition',
 'cylinders',
 'fuel',
 'odometer',
 'transmission',
 'type',
 'paint_color',
 'is_4wd',
 'date_posted',
 'days_listed']

In [13]:
# Histogram for Model Year
import plotly.express as px

# Assuming df is your DataFrame
fig = px.histogram(vehicles_dropped, x='model_year', nbins=20, title='Distribution of Model Year')
fig.show()

In [26]:
# Histogram for Type
import plotly.express as px

# Assuming df is your DataFrame
fig = px.histogram(vehicles_dropped, x='type', nbins=20, title='Distribution of Model Type')
fig.show()

In [23]:
import plotly.express as px

# Scatter plot of Cylinders versus Odometer
fig = px.scatter(vehicles_dropped, 
                 x='odometer', 
                 y='price', 
                 color='condition',  # Change this to your chosen categorical variable
                 color_discrete_sequence=['red', 'blue', 'green', 'purple', 'yellow', 'orange'],
                 title='Scatter Plot of Price vs. Odometer',
                 labels={'odometer': 'Odometer (miles)', 'price': 'Price ($)'},
                 hover_name='model')  # Optional: Shows the model on hover
fig.show()


In [27]:

# Create the scatter plot
fig = px.scatter(
    vehicles_dropped,
    x='odometer', 
    y='price', 
    color='condition',          # Color by condition
    size='model_year',          # Size of dots by model year
    hover_name='fuel',          # Show fuel type on hover
    title='Unique Scatter Plot of Price vs. Odometer',
    labels={'odometer': 'Odometer (miles)', 'price': 'Price ($)', 'model_year': 'Model Year'},
    size_max=20,                # Maximum size of bubbles
    template='plotly_dark'      # Dark theme for uniqueness
)

# Adding trendline for additional insights
fig.add_traces(px.scatter(vehicles_dropped, x='odometer', y='price', trendline='ols').data)

# Show the plot
fig.show()